### PR TITLE
Automate Versioning, Release, and PyPI Publishing

### PR DESCRIPTION
**Pull request by fredgrub**
# Pull Request Description

This pull request introduces a Git hook and a GitHub actions workflow to automate versioning, release creation, and PyPI publishing. I have closed the previous PR, https://github.com/bndr/pipreqs/pull/411, leaving it for testing. If this PR passes, we will need to close the https://github.com/bndr/pipreqs/pull/408 PR as well. Here's a summary of the changes:

1. **GitHub Actions Workflow:**
   - A workflow is triggered when a new version tag (formatted as `vX.X.X`) is pushed.
   - It automates the creation of a GitHub release and the publishing of a Python package to PyPI.

2. **Git Hook:**
   - Added a pre-commit Git hook to extract the version from `pyproject.toml` and create a corresponding version tag.

3. **Poetry bumpversion plugin**
   - Add support for https://pypi.org/project/poetry-bumpversion/. This solves the issue https://github.com/bndr/pipreqs/issues/422 in a poetry-compatible way.

# Testing this PR

- [ ] Configure `./git-hooks/` directory so that it becomes visible to git hooks.
  - `git config core.hooksPath .git-hooks`.
- [ ] Install **poetry-bumpversion** plugin.
  - `poetry self add poetry-bumpversion`.
- [ ] Update package version with `poetry version <major|minor|...>` (this will change both `pyproject.toml` and `pipreqs/__init__.py`).
- [ ] Add and commit changes, including `pyproject.toml` and `pipreqs/__init__.py`.
  - This will trigger the post-commit hook.
- [ ] Push the commit and the newly created tag.
  - For pushing a specific tag, `git push <remote> <tag>`.
- [ ] Wait for the workflow termination.

# Additional Notes
- In the GitHub action secrets setting, make sure that `PYPI_TOKEN` exists.
- For testing purposes, **you should** configure the workflow to publish to the test PyPI repository. Just add the following lines after `with:` on line 30.
  ```
  repository_name: "test-pypi"
  repository_url: "https://test.pypi.org/legacy/"
  ```
  - It's needed to create an account there. You might also want to change the token secret for `TEST_PYPI_TOKEN` (make sure that you create it on Test PyPI and added it on GitHub Action secrets).
  - You can also delete github releases and tags. If any was created by testing this workflow, just delete them.
- A successefully execution of this workflow can be found at https://github.com/pipreqsxp/pipreqs/actions/runs/7064783492. The release can be found at https://test.pypi.org/project/pipreqs-build-test/.

### CODE REVIEW

**Review by fredgrub — COMMENTED** [review]
**TL;DR: this post-commit hook ensures that valid versions are automatically tagged based on changes in the `pyproject.toml` file.**

This post-commit hook automates the tagging of the project based on changes in the `pyproject.toml` file. Here's a breakdown of what it does:

**1. Extracting the version number:**

* `git diff HEAD^..HEAD`: This line compares the current commit (`HEAD`) to its immediate predecessor (`HEAD^`).
* `-- "$(git rev-parse --show-toplevel)"/pyproject.toml`: This specifies the `pyproject.toml` file within the project root directory.
* `grep -m 1 '^\+.*version'`: This searches for the first line starting with a "+" and containing the word "version".
* `sed -s 's/[^A-Z0-9\.\-]//g'`: This removes any characters except numbers, letters, dots, and hyphens from the matched line.
* `version=`: Finally, the extracted version string is stored in the `version` variable.

**2. Validating the version format:**

* `if [[ ! $version =~ ^([0-9]+)\.([0-9]+)\.([0-9]+)(\-[A-Z]+\.[0-9]+)?$ ]]; then`: This checks if the extracted version string matches a specific format:
    * `^`: Starts with the beginning of the string.
    * `([0-9]+)`: Matches one or more digits (major version).
    * `\.`: Matches a literal dot.
    * `([0-9]+)`: Matches one or more digits (minor version).
    * `\.`: Matches a literal dot.
    * `([0-9]+)`: Matches one or more digits (patch version).
    * `(\-[A-Z]+\.[0-9]+)?`: This is optional and matches a hyphen followed by one or more uppercase letters and a dot and another number (pre-release + build information).
    * `$`: Matches the end of the string.
* If the format is invalid, it logs a message and exits with an error code.

**3. Creating the tag:**

* `git tag -a "v$version"`: This creates a new annotated tag named `v$version` (with the prefix "v") using the extracted version number.
* ``-m "`git log -1 --format=%s`"``: This sets the tag message with the full commit message of the current commit.
* `echo "Created a new tag, v$version"`: This prints a confirmation message.

File: .git-hooks/post-commit
Comment (fredgrub) [inline .git-hooks/post-commit:1]: **TL;DR: this post-commit hook ensures that valid versions are automatically tagged based on changes in the `pyproject.toml` file.**

This post-commit hook automates the tagging of the project based on changes in the `pyproject.toml` file. Here's a breakdown of what it does:

**1. Extracting the version number:**

* `git diff HEAD^..HEAD`: This line compares the current commit (`HEAD`) to its immediate predecessor (`HEAD^`).
* `-- "$(git rev-parse --show-toplevel)"/pyproject.toml`: This specifies the `pyproject.toml` file within the project root directory.
* `grep -m 1 '^\+.*version'`: This searches for the first line starting with a "+" and containing the word "version".
* `sed -s 's/[^A-Z0-9\.\-]//g'`: This removes any characters except numbers, letters, dots, and hyphens from the matched line.
* `version=`: Finally, the extracted version string is stored in the `version` variable.

**2. Validating the version format:**

* `if [[ ! $version =~ ^([0-9]+)\.([0-9]+)\.([0-9]+)(\-[A-Z]+\.[0-9]+)?$ ]]; then`: This checks if the extracted version string matches a specific format:
    * `^`: Starts with the beginning of the string.
    * `([0-9]+)`: Matches one or more digits (major version).
    * `\.`: Matches a literal dot.
    * `([0-9]+)`: Matches one or more digits (minor version).
    * `\.`: Matches a literal dot.
    * `([0-9]+)`: Matches one or more digits (patch version).
    * `(\-[A-Z]+\.[0-9]+)?`: This is optional and matches a hyphen followed by one or more uppercase letters and a dot and another number (pre-release + build information).
    * `$`: Matches the end of the string.
* If the format is invalid, it logs a message and exits with an error code.

**3. Creating the tag:**

* `git tag -a "v$version"`: This creates a new annotated tag named `v$version` (with the prefix "v") using the extracted version number.
* ``-m "`git log -1 --format=%s`"``: This sets the tag message with the full commit message of the current commit.
* `echo "Created a new tag, v$version"`: This prints a confirmation message.

File: .github/workflows/release_and_publish.yml
Comment (fredgrub) [inline .github/workflows/release_and_publish.yml:1]: This GitHub Actions workflow automates two tasks: 

1. **Create Release:**
   - Triggered when a new version tag (formatted as `vX.X.X`) is pushed.
   - Creates a release with the tag name, marking it as the latest version.

2. **Publish to PyPI:**
   - Runs after the release is created successfully.
   - Builds and publishes the Python package to PyPI using Poetry.
   - Ignores development requirements during the publishing process.

This ensures that each new version gets a release on GitHub and the corresponding Python package is published on PyPI.

File: .git-hooks/post-commit
Comment (alan-barzilay) [reply .git-hooks/post-commit:1]: its dependecies are bash grep and sed, right? Maybe we should add this somewhere, maybe add to the contribute section a tutorial for the hook installation (like the one you created for this pr) with a note about dependecies

**Comment by alan-barzilay** [comment]
>  Install poetry-bumpversion plugin.
> 
>     poetry self add poetry-bumpversion.
> 

isnt there a way to set this as a dev dependency?

**Comment by fredgrub** [comment]
> >  Install poetry-bumpversion plugin.
> > 
> >     poetry self add poetry-bumpversion.
> > 
> 
> isn't there a way to set this as a dev dependency? Since it is a pipy package

Yeah! I searched about it and found that it is possible. I found that using `self add` is a way of installing the plug-in globally. If we declare the plug-in as dev dependency, it will be available only to the target project.

I'll change it ASAP.

File: .git-hooks/post-commit
Comment (alan-barzilay) [reply .git-hooks/post-commit:1]: please add this breakdown of the logic to the commit message with git rebase reword

File: .github/workflows/release_and_publish.yml
Comment (alan-barzilay) [reply .github/workflows/release_and_publish.yml:1]: Same thing, please add this breakdown of the logic to the commit message with git rebase reword. If you made such a good explanation there is no reason to leave it out of the commit :)

**Comment by fredgrub** [comment]
> ```
> git config core.hooksPath .git-hooks
> ```
> shouldnt we also add --local? I may be misunderstanding, but without the local flag, the hooks will be available in all git projects system wide

According to `git-config` documentation `--local` is the default option.

**Review by alan-barzilay — APPROVED** [review]
everything looks good to me! I will not merge it yet due to the secret not being set yet

**Comment by alan-barzilay** [comment]
Since I have no permission to set the secret, I have been reaching out to @bndr to set it but I've got no response for the last couple months. Maybe he is busy, but this wont be merged until the secret is set so sadly this pr might be dead. I truly appreciate all the work put into it and apologize for this, I probably should have explored this issue before allowing work to be done on this

**Comment by jonas-eschle** [comment]
Hi @alan-barzilay getting back to this, is this still up to date? I think we can start thinking about merging it!

**Comment by alan-barzilay** [comment]
Hi @jonas-eschle,

I saw I was tagged in several pull requests today, and instead of responding to each one individually, I thought it might be better to write here to give some broader context — both about my involvement with the project and some of the structural challenges that remain.

First, thanks for stepping in and contributing. I hadn't seen your name in earlier activity, but it looks like you've recently taken on a more active role. Were you added as a maintainer by @bndr? And are you in contact with him?

For some context: I was previously added as a maintainer, eventually I gathered a team and we worked on a few releases of pipreqs. During that time, we handled a lot of backlog and implemented some long-requested features — including support for Jupyter notebooks, which had been a consistent ask from the community since early on. While doing that work, I came up against some limitations: as a maintainer (but not the repo owner), I couldn't manage contributor access or configure things like repository secrets. I reached out to @bndr about potentially creating a GitHub organization or transferring ownership, but I never got a response regarding this issue or any of the others I raised, which made it challenging to keep the project moving forward in a sustainable way.
Without the ability to manage the project effectively, and considering some personal commitments that I had to take care of, I eventually decided to step away.

At this point, I’m no longer involved with the project and don’t plan to return to active maintenance given its current state. That said, I’m glad to see new activity, and I do hope the project continues to evolve. If there’s ever interest in transitioning pipreqs to a GitHub organization or if you’d like to talk about ways to improve the project or its long-term sustainability, I’d be happy to chat. I can provide a throwaway email if you'd prefer to connect directly and streamline that discussion.

Thanks again for stepping in and helping keep the project going. Best of luck moving things forward!

**Comment by alan-barzilay** [comment]
> Hi @alan-barzilay getting back to this, is this still up to date? I think we can start thinking about merging it!

iirc this will fail and can't move forward without setting a repo secret, which only the owner can do

**Comment by jonas-eschle** [comment]
Hi @alan-barzilay ,

> I saw I was tagged in several pull requests today, and instead of responding to each one individually, I thought it might be better to write here to give some broader context — both about my involvement with the project and some of the structural challenges that remain.

First of all, thanks for all the work you've done so far for the repo already

> Were you added as a maintainer by @bndr? And are you in contact with him?
yes and yes
 
> For some context: I was previously added as a maintainer
thanks a lot for the context! I have to say, I agree also with your point of moving it to an orga, it complicates things a lot. And I think there is also slightly suboptimal default behavior.

> If there’s ever interest in transitioning pipreqs to a GitHub organization or if you’d like to talk about ways to improve the project or its long-term sustainability, I’d be happy to chat. I can provide a throwaway email if you'd prefer to connect directly and streamline that discussion.

I think this could be beneficial, please just write an e-mail and we can take it from there, thanks for proposing this!

> Thanks again for stepping in and helping keep the project going. Best of luck moving things forward!
Thanks for all the work you've put in!

**Comment by alan-barzilay** [comment]
> I think this could be beneficial, please just write an e-mail and we can take it from there, thanks for proposing this!


7cxti2np3@mozmail.com
